validators/cgroup_validator_linux.go

```diff
@@ -142,15 +166,41 @@ func (c *CgroupsValidator) getCgroupV2Subsystems() ([]string, error) {
 	// Some controllers are implicitly enabled by the kernel.
 	// Those controllers do not appear in /sys/fs/cgroup/cgroup.controllers.
 	// https://github.com/torvalds/linux/blob/v5.3/kernel/cgroup/cgroup.c#L433-L434
-	// We assume these are always available, as it is hard to detect availability.
-	// So, we hardcode the following as "pseudo" controllers.
+	// For freezer, we use checkCgroupV2Freeze() to check.
+	// For others, we assume these are always available, as it is hard to detect availability.
+	// We hardcode the following as initial controllers.
 	// - devices: implemented in kernel 4.15
-	// - freezer: implemented in kernel 5.2
-	pseudo := []string{"devices", "freezer"}
+	subsystems := []string{"devices"}
+	freezeSupported, err := checkCgroupV2Freeze()
+	if err != nil {
+		fmt.Printf("error checking cgroups v2 freeze: %v\n", err)
+	}
+	if freezeSupported {
+		subsystems = append(subsystems, "freezer")
+	}
 	data, err := ioutil.ReadFile(filepath.Join(unifiedMountpoint, "cgroup.controllers"))
 	if err != nil {
 		return nil, err
 	}
-	subsystems := append(pseudo, strings.Fields(string(data))...)
+	subsystems = append(subsystems, strings.Fields(string(data))...)
 	return subsystems, nil
 }
+
+// To check if the freezer controller is enabled in Linux kernels 5.2, we can check the existence of cgroup.freeze.
+func checkCgroupV2Freeze() (bool, error) {
+	tmpDir, err := os.MkdirTemp(unifiedMountpoint, "freezer-test")
+	if err != nil {
+		return false, err
+	}
+	defer func() {
+		err := os.RemoveAll(tmpDir)
+		if err != nil {
+			fmt.Printf("error removing directory %q: %v\n", tmpDir, err)
+		}
+	}()
+	_, err = os.Stat(filepath.Join(tmpDir, "cgroup.freeze"))
+	if err == nil {
```

Comment:
Can we use os.IsNotExist(err) to clarify the root cause?
Not the absence of the cgroup.freeze but other errors may happen.

Reply:
Updated.
- Should I print error if it is not `NotExistErr`?